docs/content/docs/dev/datastream/dataset_migration.md

```diff
@@ -0,0 +1,758 @@
+        .apply((leftIterable, rightInterable, collector) -> {
+            if(!rightInterable.iterator().hasNext()){
+            // implement user-defined left outer join logic
+            }
+        });
+// right outer join
+dataStream1.coGroup(dataStream2)
+        .where(value -> value.f0)
+        .equalTo(value -> value.f0)
+        .window(GlobalWindows.create())
+        .trigger(new EOFTrigger())
+        .apply((leftIterable, rightInterable, collector) -> {
+            if(!leftIterable.iterator().hasNext()){
+            // implement user-defined right outer join logic
+            }
+        });
+        </code>
+    </td>
+  </tr>
+</table>
+
+#### Category 3
+
+For category 3, these DataSet APIs can be implemented by DataStream APIs with different semantic and different calculation behavior. Additional 
+calculation steps will be added.
+
+To collect records from each subtask, every record needs to be assigned a unique subtask ID and grouped accordingly within the window. 
+This additional step of assigning the subtask ID and performing a groupby operation introduces shuffle costs. Here is an example code 
+snippet showing how to assign a subtask ID to each record. The function assignSubtaskID is used to explain the detailed behavior and will 
+be utilized in the subsequent sections:
+```java
+// assign subtask ID to all records
+DataStream<Tuple2<String, Integer>> assignSubtaskID(DataStream<Integer> dataStream) {
+        return dataStream.map(new RichMapFunction<Integer, Tuple2<String, Integer>>() {
+                @Override
+                public Tuple2<String, Integer> map(Integer value) {
+                        return Tuple2.of(String.valueOf(getRuntimeContext().getIndexOfThisSubtask()), value);
+                }
+        });
+}
+```
+
+### MapPartition/SortPartition
+
+<table>
+  <tr>
+    <th>DataSet</th>
+    <th>DataStream</th>
+  </tr>
+  <tr>
+    <td>
+        <code style="white-space: pre-line;">DataSet&ltInteger&gt dataSet = // [...]
+// MapPartition
+dataSet.mapPartition(new MapPartitionFunction(){
+        // implement user-defined map partition logic
+        });
+// SortPartition
+dataSet.sortPartition(0, Order.ASCENDING);
+dataSet.sortPartition(0, Order.DESCENDING)
+        </code>
+    </td>
+    <td>
+        <code style="white-space: pre-line;">DataStream&ltInteger&gt dataStream = // [...]
+// assign subtask ID to all records
+DataStream&ltTuple2&ltString, Integer&gt&gt dataStream1 = assignSubtaskID(dataStream);
+dataStream1.keyBy(value -> value.f0)
+        .window(GlobalWindows.create())
+        .trigger(new EOFTrigger())
+        .apply(new WindowFunction(){
+        // implement user-defined map partition or sort partition logic
+        });
+        </code>
+    </td>
+  </tr>
+</table>
+
+### Cross
+
+To implement the Cross operation between two DataStreams, datastream1 and datastream2, the parallism of two datastreams should be the
+same and then follow these steps:
```

Comment:
The `Cross`  builds all pairwise combinations of the elements of both input DataSets. Here is the topology diagram of `dataSet1.cross(dataSet2)`, the parallelism `dataSet1` is different from `dataSet2`.
![image](https://github.com/apache/flink/assets/104502720/7fc09ebc-93ae-4550-9525-2bd8442d2317)


If DataStream wants to achieve `Cross`, there are two options:

Option1: `dataStream1.join(dataStream2).where(samekey).equalTo(samekey)`
This will make all data are processed in a single subtask after crossed and result in single parallelism.

Option2: 
1. Broadcast `dataStream1` in several subtasks and make each subtask contain all data of `dataStream1`.
2. Set same parallelism to `dataStream1` and `dataStream2`.
3. Join `dataStream1` with `dataStream2` according to the subtask id. Then each subtask of `dataStream2` can obtain all data of `dataStream1`, which could complete the calculation of `Cross`.

Reply:
I wrote a program like follows according to the document. The expected result size is 100 while the actual result is not. Could you please help verify the correctness of the following code and the document?

```java
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
env.setParallelism(4);
DataStream<Long> dataStream1 = env.fromSequence(0, 9);
DataStream<Long> dataStream2 = env.fromSequence(0, 9);

DataStream<Tuple2<String, Long>> dataStream3 = dataStream1.broadcast().map(new AddSubtaskIDMapFunction<>());
DataStream<Tuple2<String, Long>> dataStream4 = dataStream2.map(new AddSubtaskIDMapFunction<>());
// join the two streams according to the subtask ID
DataStream<Long> resultStream = dataStream3.join(dataStream4)
        .where(value -> value.f0)
        .equalTo(value -> value.f0)
        .window(GlobalWindows.create())
        .trigger(new EOFTrigger())
        .apply((JoinFunction<Tuple2<String, Long>, Tuple2<String, Long>, Long>) (first, second) ->
                first.f1 * 10 + second.f1);

List<Long> result = IteratorUtils.toList(resultStream.executeAndCollect());
System.out.println(result.size());
Collections.sort(result);
System.out.println(result);
```